File: problems/0074-search-a-2d-matrix/typescript/solution.ts
// Search a 2D Matrix — https://leetcode.com/problems/search-a-2d-matrix/
export function searchMatrix(matrix: number[][], target: number): boolean {
  if (matrix.length === 0 || matrix[0].length === 0) return false;
  const rows = matrix.length;
  const cols = matrix[0].length;
  let lo = 0;
  let hi = rows * cols - 1;
  while (lo <= hi) {
    const mid = Math.floor((lo + hi) / 2);
    const val = matrix[Math.floor(mid / cols)][mid % cols];
    if (val === target) return true;
    if (val < target) lo = mid + 1;
    else hi = mid - 1;
  }
  return false;
}

File: problems/0074-search-a-2d-matrix/go/solution.go
package solution

// Search a 2D Matrix — https://leetcode.com/problems/search-a-2d-matrix/
func searchMatrix(matrix [][]int, target int) bool {
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return false
	}
	rows, cols := len(matrix), len(matrix[0])
	lo, hi := 0, rows*cols-1
	for lo <= hi {
		mid := lo + (hi-lo)/2
		val := matrix[mid/cols][mid%cols]
		if val == target {
			return true
		}
		if val < target {
			lo = mid + 1
		} else {
			hi = mid - 1
		}
	}
	return false
}
